src/Microsoft.Diagnostics.Runtime/src/Builders/RuntimeBuilder.cs

```diff
@@ -1002,6 +1012,136 @@ public ClrType CreateSystemType(ClrHeap heap, ulong mt, string typeName)
+        {
+            // ECMA 335 - II.23.2.12 - Type
+
+            if (!parser.GetElemType(out ClrElementType etype))
+                return null;
+
+            if (etype.IsPrimitive() || etype == ClrElementType.Void || etype == ClrElementType.Object || etype == ClrElementType.String)
+                return GetOrCreateBasicType(etype);
+
+            if (etype == ClrElementType.Array)
+            {
+                ClrType? innerType = GetOrCreateTypeFromSignature(module, parser, typeParameters, methodParameters);
+                innerType ??= GetOrCreateBasicType(ClrElementType.Void);  // Need a placeholder if we can't determine type
+
+                // II.23.2.13
+                if (!parser.GetData(out int rank))
+                    return null;
+
+                if (!parser.GetData(out int numSizes))
+                    return null;
+
+                for (int i = 0; i < numSizes; i++)
+                    if (!parser.GetData(out _))
+                        return null;
+
+                if (!parser.GetData(out int numLowBounds))
+                    return null;
+
+                for (int i = 0; i < numLowBounds; i++)
+                    if (!parser.GetData(out _))
+                        return null;
+
+                // We should probably use sizes and lower bounds, but this is so rare I won't worry about it for now
+                ClrType? result = GetOrCreateArrayType(innerType, rank);
+                return result;
+            }
+
+            if (etype == ClrElementType.Class || etype == ClrElementType.Struct)
+            {
+                if (!parser.GetToken(out int token))
+                    return null;
+
+                ClrType? result = module != null ? GetOrCreateTypeFromToken(module, token) : null;
+                if (result == null)
+                {
+                    // todo, create a type from metadata instead of returning a basic type?
+                    result = GetOrCreateBasicType(etype);
+                }
+
+                return result;
+            }
+
+            if (etype == ClrElementType.FunctionPointer)
+            {
+                if (!parser.GetToken(out _))
+                    return null;
+
+                // We don't have a type for function pointers so we'll make it a void pointer
+                ClrType inner = GetOrCreateBasicType(ClrElementType.Void);
+                return GetOrCreatePointerType(inner, 1);
+            }
+
+            if (etype == ClrElementType.GenericInstantiation)
+            {
+                if (!parser.GetElemType(out ClrElementType classOrStruct))
+                    return null;
+
+                if (!parser.GetToken(out int token))
+                    return null;
+
+
+                if (!parser.GetData(out int count))
+                    return null;
+
+                var builder = ImmutableArray.CreateBuilder<ClrType>(count);
```

Comment:
todo: Remove this builder.  We still need to fully walk the types here so that SigParser is left in a state pointing after the generic instantiation, but we don't need to build an immutable array since we don't use it.